Archipelago.HollowKnight/Archipelago.cs

```diff
@@ -182,17 +194,141 @@ private void ModHooks_HeroUpdateHook()
             {
                 ReceiveNextItem();
             }
+
+            if (deathlinkStatus == DeathLinkStatus.Pending && CanMurderPlayer())
+            {
+                MurderPlayer();
+            }
+        }
+
+        public bool CanMurderPlayer()
+        {
+            HeroController hc = HeroController.instance;
+            return hc.acceptingInput && hc.damageMode == GlobalEnums.DamageMode.FULL_DAMAGE;
+        }
+
+        public void MurderPlayer()
+        {
+            HeroController hc = HeroController.instance;
+            string scene = GameManager.instance.sceneName;
+            LogDebug($"So, somebody else has chosen... death.  Current scene: {scene}");
+
+            bool switcharoo = (!(
+                // These are all conditions where the death should be processed like a vanilla death with no shade switcharoo
+                SlotOptions.DeathLink == DeathLinkType.Vanilla
+                || (SlotOptions.DeathLink == DeathLinkType.Shade && PlayerData.instance.shadeScene == "None")
+                || (scene.StartsWith("Dream_") && scene != SceneNames.Dream_Final_Boss)  // Most dream sequences, except Radiance
+                || scene.StartsWith("GG_")  // All Godhome sequences
+                || scene.StartsWith("White_Palace_")  // White Palace is a dream.
+                || scene == SceneNames.Grimm_Nightmare  // NKG
+            ));
+
+            if (switcharoo && scene == "Room_Colosseum_Spectate")
```

Comment:
So there's a chance to overwrite a shade if you set it to shadeless deathlink and die in room_colosseum_spectate?

Reply:
Whoops!  This logic is actually backwards, it should _never_ produce a shade in that room.  (The shade spawns out of bounds -- it's a vanilla bug but normally the player can never die there.)

I _think_ we also need this exception for the Lifeblood Core room because AFAIK shade spawns are out of bounds there.

Realistically, better longterm fixes might be letting those two scenes be handled normally, detecting shades in those two areas and explicitly moving them to reasonable spawn locations.